heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type * as TWebpack from 'webpack';
+import type TWebpackDevServer from 'webpack-dev-server';
+import { AsyncParallelHook, AsyncSeriesBailHook, AsyncSeriesHook } from 'tapable';
+import { FileError, LegacyAdapters } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  IHeftTaskSession,
+  IHeftTaskCleanHookOptions,
+  IHeftTaskPlugin,
+  IHeftTaskRunHookOptions,
+  IScopedLogger
+} from '@rushstack/heft';
+
+import type {
+  IWebpackConfiguration,
+  IWebpackConfigurationWithDevServer,
+  IWebpack5PluginAccessor
+} from './shared';
+import { WebpackConfigurationLoader } from './WebpackConfigurationLoader';
+
+export interface IWebpack5PluginOptions {
+  devConfigurationPath: string | undefined;
+  configurationPath: string | undefined;
+}
+
+/**
+ * @public
+ */
+export const PLUGIN_NAME: 'Webpack5Plugin' = 'Webpack5Plugin';
+const WEBPACK_DEV_SERVER_PACKAGE_NAME: 'webpack-dev-server' = 'webpack-dev-server';
+const WEBPACK_DEV_SERVER_ENV_VAR_NAME: 'WEBPACK_DEV_SERVER' = 'WEBPACK_DEV_SERVER';
+const UNINITIALIZED: 'UNINITIALIZED' = 'UNINITIALIZED';
+
+/**
+ * @internal
+ */
+export default class Webpack5Plugin implements IHeftTaskPlugin<IWebpack5PluginOptions> {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackConfiguration: IWebpackConfiguration | undefined | typeof UNINITIALIZED = UNINITIALIZED;
+
+  public readonly accessor: IWebpack5PluginAccessor = {
+    hooks: {
+      onLoadConfiguration: new AsyncSeriesBailHook(),
+      onConfigure: new AsyncSeriesHook(['webpackConfiguration']),
+      onAfterConfigure: new AsyncParallelHook(['webpackConfiguration']),
+      onEmitStats: new AsyncParallelHook(['webpackStats'])
+    }
+  };
+
+  public apply(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IWebpack5PluginOptions
+  ): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    // Provide the initial webpack configuration by tapping the hook. Tap with stage as
+    // MAX_SAFE_INTEGER to ensure that we run last.
```

Comment:
Same comment about this being a somewhat weird way to tap a hook, especially with a default.